### PR TITLE
Gauge: Simplify gauge dimension panel options

### PR DESCRIPTION
**Pull request by nmarrs**
Update the gauge panel with a similar approach taken in bar gauge panel (see [this PR](https://github.com/grafana/grafana/pull/76042)) for handling `auto` / `manual` gauge sizing options. By default `auto` size will use a reasonable hardcoded value (though in the future this could be made smarter) while `manual` allows user to explicitly determine the size of their gauge.

UX

https://github.com/grafana/grafana/assets/22381771/9e670974-a535-4c10-a1e6-9c294cf327a0

[Demo dashboard.txt](https://github.com/grafana/grafana/files/12850512/En.Gauge-1696884218759.json.txt)

Related to https://github.com/grafana/grafana/pull/71690 + https://github.com/grafana/grafana/pull/46819

For docs changes see https://github.com/grafana/grafana/pull/76240

Created a [follow-up issue for adding an overflow notice inside of VizRepeater](https://github.com/grafana/grafana/issues/77836)

### CODE REVIEW

File: public/app/plugins/panel/gauge/GaugePanel.tsx [hunk 1]
```diff
@@ -81,6 +82,21 @@ export class GaugePanel extends PureComponent<PanelProps<Options>> {
...
+    const numberOfGauges = this.getValues().length;
+
+    const remainderWidth = width - numberOfGauges * GAUGE_MIN_WIDTH;
+    const isHorizontalOverflow = options.orientation === VizOrientation.Horizontal && remainderWidth < 0;
+
+    const remainderHeight = height - numberOfGauges * GAUGE_MIN_HEIGHT;
+    const isVerticalOverflow = options.orientation === VizOrientation.Vertical && remainderHeight < 0;
+
+    const isOverflow = isHorizontalOverflow || isVerticalOverflow;
+
```
Comment (torkelo) [inline public/app/plugins/panel/gauge/GaugePanel.tsx:99]: can't this logic be inside the VizRepeater instead? it knows the number of elements, and you pass GAUGE_MIN_WIDTH  via minVizWidth (same with height), it knows the orientation etc. 

The big problem (beyond having to duplicate it for bar gauge) with having it here is needing that additional call to getValues

**Review by torkelo — COMMENTED** [review]
yes, something like this was what I had in mind. It could even be a bit smarter, if where to use alignment factors like BarGauge, then it would know the longest gauge name and value and have a better sense of size constraint (but maybe not needed for a first iteration)

File: public/app/plugins/panel/gauge/panelcfg.cue [hunk 2]
```diff
@@ -29,8 +29,9 @@ composableKinds: PanelCfg: {
 					common.SingleStatBaseOptions
 					showThresholdLabels:  bool | *false
 					showThresholdMarkers: bool | *true
-					minVizWidth:          uint32 | *75
-					minVizHeight:         uint32 | *75
+					sizing:               common.BarGaugeSizing & (*"auto" | _)
```
Comment (nmarrs) [inline public/app/plugins/panel/gauge/panelcfg.cue:32]: We can rename the `BarGaugeSizing` to just `GaugeSizing` to make it more universal if desired - I don't feel too strongly on need to do this 🤷‍♂️

**Comment by nmarrs** [comment]
@torkelo I've updated this PR to have the same handling of gauge dimensions as [recent bar gauge PR](https://github.com/grafana/grafana/pull/76042) (`Auto` / `Manual` sizing option that hides extra panel options by default)

File: public/app/plugins/panel/gauge/module.tsx [hunk 3]
```diff
@@ -39,19 +39,33 @@ export const plugin = new PanelPlugin<Options>(GaugePanel)
...
+        settings: {
+          options: [
+            { value: BarGaugeSizing.Auto, label: 'Auto' },
+            { value: BarGaugeSizing.Manual, label: 'Manual' },
+          ],
+        },
+        defaultValue: defaultOptions.sizing,
+        showIf: (options: Options) => options.orientation !== VizOrientation.Auto,
+      })
       .addNumberInput({
```
Comment (Develer) [inline public/app/plugins/panel/gauge/module.tsx:54]: Hey @nmarrs Would you mind to replace number input with slider input?
```suggestion
.addSliderInput({
        path: 'minVizWidth',
        name: 'Min width',
        description: 'Minimum column width (vertical orientation)',
        defaultValue: defaultOptions.minVizWidth,
        settings: {
          min: 0,
          max: 300,
          step: 1,
        },
        showIf: (options: Options) =>
          options.sizing === BarGaugeSizing.Manual && options.orientation === VizOrientation.Vertical,
      })
      .addSliderInput({
        path: 'minVizHeight',
        name: 'Min height',
        description: 'Minimum row height (horizontal orientation)',
        defaultValue: defaultOptions.minVizHeight,
        settings: {
          min: 0,
          max: 300,
          step: 1,
        },
        showIf: (options: Options) =>
          options.sizing === BarGaugeSizing.Manual && options.orientation === VizOrientation.Horizontal,
      });
```

**Review by Develer — APPROVED** [review]

File: public/app/plugins/panel/gauge/module.tsx (hunk 3 above)
Comment (Develer) [reply public/app/plugins/panel/gauge/module.tsx:54]: please, adjust max value as needed 🙂

Comment (Develer) [reply public/app/plugins/panel/gauge/module.tsx:54]: 600 max value should be more than enough. In other cases, if e.g. user wants a bigger gauge - auto move should cover that. Let me know what you think.

**Comment by Develer** [comment]
@nmarrs btw, just noticed, shouldn't `Gauge size` show also in auto layout orientation mode?

**Comment by nmarrs** [comment]
> @nmarrs btw, just noticed, shouldn't `Gauge size` show also in auto layout orientation mode?

@Develer for the gauge panel `auto` layout behaves differently from the bar gauge panel (no need for overflow handling as it lays out gauges neatly in available space) so including the gauge size option in that case doesn't really make as much sense IMO (nor do the gauge size options have any impact on gauge behavior when in auto mode (in the code))
Screenshot 2023-11-17 at 11 31 24 AM

File: public/app/plugins/panel/gauge/module.tsx (hunk 3 above)
Comment (nmarrs) [reply public/app/plugins/panel/gauge/module.tsx:54]: checking it out locally now :) goal is to get this merged today (and associated docs PR) 👍

Comment (nmarrs) [reply public/app/plugins/panel/gauge/module.tsx]: 600 seems reasonable enough 👍 updated PR with requested changes

File: public/app/plugins/panel/gauge/panelcfg.gen.ts [hunk 4]
```diff
@@ -15,11 +15,13 @@ export interface Options extends common.SingleStatBaseOptions {
...
   showThresholdLabels: boolean;
   showThresholdMarkers: boolean;
+  sizing: common.BarGaugeSizing;
 }
 
 export const defaultOptions: Partial<Options> = {
-  minVizHeight: 75,
-  minVizWidth: 75,
+  minVizHeight: 200,
+  minVizWidth: 200,
```
Comment (torkelo) [inline public/app/plugins/panel/gauge/panelcfg.gen.ts:23]: @nmarrs this is a pretty big breaking change, radically changing the look of thousands of panels that used to show many gauges now just show a few and with a scrollbar.

**Comment by melroy89** [comment]
I bet I got regression on this PR .....

image

File: public/app/plugins/panel/gauge/panelcfg.gen.ts (hunk 4 above)
Comment (melroy89) [reply public/app/plugins/panel/gauge/panelcfg.gen.ts:23]: You can get it causes issues.. Not cool.

**Pull request merged**